middle-pgsql.cpp

```diff
@@ -1072,7 +1072,10 @@ void middle_pgsql_t::pgsql_stop_one(table_desc *table)
 void middle_pgsql_t::stop(void)
 {
     cache.reset();
-    if (out_options->flat_node_cache_enabled) persistent_cache.reset();
+    if (out_options->flat_node_cache_enabled) {
+        persistent_cache->clean_up();
+        persistent_cache.reset();
+    }
```

Comment:
I still think that the solution should be generalized, so the benefit is there even without --drop. To achieve that the stop() function needs to be split at this point, the upper part executed before the indexing and the lower part needs to remain where the stop() call was before this change.

And if we do that, then the future vector below can even be merged with the one with the indexing functions in osmdata.cpp below, meaning that we get a much better parallelization.

Reply:
From what I understand, there are 5 modes of import:

1. Normal mode, where intermediate data is created in RAM. It's not clear to me from the doc if this mode is updatable or not, but I guess the answer is yes.
2. `--slim` mode, where that data is written in the database. This mode is updatable.
3. `--slim` + `--flat-nodes`, where par of the data if written in a spaces file and `mmap()`'ed, so no db queries. This mode is also updatable.
4. `--slim --drop` and
5. `--slim --flat-nodes --drop`, were the intermediate data is erased. With this flag, the database is not upgradable anymore.

There are two problems with `--slim --drop`:

* The flat nodes file, if used (mode 5), is not erased. This PR fixes this.
* The intermediate data is removed at the end, even when they're not longer used way before that moment (mainly, just before indexing). This PR fixes this only for `--flat-nodes`, mode 5.

So what you mean is that we should do the second item for both modes 5 and 3. I think it makes sense, and it seems that what you mention about splitting `stop()` is actually making this slower to run: I'm getting 500s+ with this patch vs 230s+ without for importing Argentina in my machine.

I'll give the `stop()` splitting a try tomorrow, but you will have to explain a little bit more about the last paragraph. Can I drop by any IRC channel to discuss it? I think I'll be faster than just ping-pong through GitHub.

Reply:
The "normal mode" (your #1) is not updateable. Updating requires intermediate data to be persistent.

Reply:
In all modes a cache of node locations is kept in memory (the -C parameter). Currently it is kept around in the indexing phase, making the memory unavailable for postgres. But I think you are right, just splitting as I suggested again leaves the drop case in a worse state. This clearly needs some more thinking. So I would suggest to leave it for this PR and tackle the cache dropping later.